File: csharp/demo/demo15a-deadlock.cs
/*
 * DEADLOCK
 * Version A
 */
using System;
using System.Threading;



class Demo15A : IRunnable
{
    private Mutex mut;


    public void run()
    {
        mut = new Mutex();

        var thFoo = new Thread(() => doTask("foo"));
        var thBar = new Thread(() => doTask("bar"));

        thFoo.Start();
        thBar.Start();

        thFoo.Join();
        thBar.Join();

        // The app may throw System.Threading.AbandonedMutexException
        Console.WriteLine("You will never see this statement due to deadlock!");
    }


    private void doTask(string name)
    {
        mut.WaitOne();

        Console.WriteLine(name + " acquired resource");

        // mut.ReleaseMutex(); // forget this statement ==> deadlock
    }
}

File: cpp-std/demo15a-deadlock.cpp
/*
DEADLOCK
Version A
*/


#include <iostream>
#include <string>
#include <thread>
#include <mutex>
using namespace std;



std::mutex mut;



void routine(std::string name) {
    mut.lock();

    cout << name << " acquired resource" << endl;

    // mut.unlock(); // forget this statement ==> deadlock
}



int main() {
    auto thFoo = std::thread(routine, "foo");
    auto thBar = std::thread(routine, "bar");

    thFoo.join();
    thBar.join();

    cout << "You will never see this statement due to deadlock!" << endl;
    return 0;
}
